synapse/rest/client/v2_alpha/account.py

```diff
@@ -158,7 +159,64 @@ def __init__(self, hs):
             hs (synapse.server.HomeServer): server
         """
         super(PasswordResetSubmitTokenServlet, self).__init__()
-        self.hs = hs
+        self.config = hs.config
+        if self.config.threepid_behaviour_email == ThreepidBehaviour.LOCAL:
+            (self.confirmation_email_template,) = load_jinja2_templates(
+                self.config.email_template_dir,
+                [self.config.email_password_reset_template_confirmation_html],
+            )
+
+    async def on_GET(self, request, medium):
+        # We currently only handle threepid token submissions for email
+        if medium != "email":
+            raise SynapseError(
+                400, "This medium is currently not supported for password resets"
+            )
+        if self.config.threepid_behaviour_email == ThreepidBehaviour.OFF:
+            if self.config.local_threepid_handling_disabled_due_to_email_config:
+                logger.warning(
+                    "Password reset emails have been disabled due to lack of an email config"
+                )
+            raise SynapseError(
+                400, "Email-based password resets are disabled on this server"
+            )
+
+        sid = parse_string(request, "sid", required=True)
+        token = parse_string(request, "token", required=True)
+        client_secret = parse_string(request, "client_secret", required=True)
+
+        # Show a confirmation page, just in case someone accidentally clicked this link when
+        # they didn't mean to
+        template_vars = {
+            "sid": sid,
+            "token": token,
+            "client_secret": client_secret,
+            "medium": medium,
+        }
+        respond_with_html(
+            request, 200, self.confirmation_email_template.render(**template_vars)
+        )
+
+
+class PasswordResetConfirmationSubmitTokenServlet(RestServlet):
+    """Handles confirmation of 3PID validation token submission.
+
+    A user will land on PasswordResetSubmitTokenServlet, confirm the password reset, then
+    submit the same parameters to this servlet.
+    """
+
+    PATTERNS = [
+        re.compile(
+            "^/_synapse/client/account/v1/password_reset/(?P<medium>[^/]*)/submit_token_confirm$"
```

Comment:
feels like we should probably hardcode `email` here, until someone invents a different medium for us to support.

Reply:
I was indeed not sure whether to stick `email` here or to make it future-proof.

But seeing as we only support `email` atm, it probably is more confusing to keep the flexibility at the moment. I'll take it out.

Reply:
I'll leave the jinja `{{ medium }}` variable in the template file though so people won't need to update their templates in the future if more mediums become available.